torchqdynamics/adaptive.py

```diff
@@ -0,0 +1,166 @@
+
+    @torch.no_grad()
+    def update_tstep(self, dt, error):
+        """Update the time step of an adaptive step size solver.
+
+        See Equation (4.12) and (4.13) of `Hairer et al., Solving Ordinary Differential
+        Equations I (1993), Springer Series in Computational Mathematics` for the
+        detailed steps.
+        """
+        if error == 0:  # no error -> maximally increase the time step
+            return dt * self.max_factor
+
+        # optimal time step
+        dt_opt = dt * error**(-1.0 / self.order)
+
+        if error <= 1:  # time step accepted -> take next time step at least as large
+            return dt * min(self.max_factor, max(1.0, self.factor * dt_opt))
+
+        if error > 1:  # time step rejected -> reduce next time step
+            return dt * min(0.9, max(self.min_factor, self.factor * dt_opt))
+
+
+class DormandPrince45(AdaptiveSolver):
+    """Dormand-Prince method for adaptive time step ODE integration.
+
+    This is a fifth order solver that uses a fourth order solution to estimate the
+    integration error. It does so using only six function evaluations. See `Dormand and
+    Prince, A family of embedded Runge-Kutta formulae (1980), Journal of Computational
+    and Applied Mathematics`.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.order = 5
+        self.tableau = self.build_tableau()
+
+    def build_tableau(self):
+        """Build the Butcher tableau of the integrator."""
+        alpha = [1 / 5, 3 / 10, 4 / 5, 8 / 9, 1.0, 1.0, 0.0]
+        beta = [
+            [1 / 5, 0, 0, 0, 0, 0, 0],
+            [3 / 40, 9 / 40, 0, 0, 0, 0, 0],
+            [44 / 45, -56 / 15, 32 / 9, 0, 0, 0, 0],
+            [19372 / 6561, -25360 / 2187, 64448 / 6561, -212 / 729, 0, 0, 0],
+            [9017 / 3168, -355 / 33, 46732 / 5247, 49 / 176, -5103 / 18656, 0, 0],
+            [35 / 384, 0, 500 / 1113, 125 / 192, -2187 / 6784, 11 / 84, 0],
+        ]
+        csol = [35 / 384, 0, 500 / 1113, 125 / 192, -2187 / 6784, 11 / 84, 0]
+        cerr = [
+            1951 / 21600,
+            0,
+            22642 / 50085,
+            451 / 720,
+            -12231 / 42400,
+            649 / 6300,
+            1 / 60,
+        ]
+
+        alpha = torch.tensor(alpha, dtype=self.t_dtype, device=self.device)
+        beta = torch.tensor(beta, dtype=self.y_dtype, device=self.device)
+        csol = torch.tensor(csol, dtype=self.y_dtype, device=self.device)
+        cerr = torch.tensor(cerr, dtype=self.y_dtype, device=self.device)
+        return (alpha, beta, csol, csol - cerr)
+
+    def step(self, f0: Tensor, y0: Tensor, t0: float,
+             dt: float) -> Tuple[Tensor, Tensor, Tensor]:
+        """Compute a single step of the ODE integration."""
+        # import butcher tableau
+        alpha, beta, csol, cerr = self.tableau
+
+        # compute iterated Runge-Kutta values
+        k = torch.zeros(7, *f0.shape, dtype=f0.dtype, device=f0.device)
```

Comment:
Shouldn't we do `torch.tensor(alpha).to(t0)` instead? It would be more straightfoward than storing the types aside + currently it is possible to have actually a different type and device for `alpha` and `t0` 

Same comment for `y` and `beta` etc...

Reply:
Thanks for raising this. Actually I have two partial answers to this:

1. It's theoretically faster/cleaner to initialize the tensor directly by specifying the dtype and device, than to initialize it and then transfer it to the proper dtype/device using `.to(...)`. It avoids GPU/CPU transfers.
2. I think there's no reason for the `AdaptiveSolver` class to know about `t0` or `y0`. It should not be an attribute of the class, so I don't really want to pass it as argument.

However my first point is quite of whatever because this is a not a performance-sensitive part of the code. The only the second point is left. Do you think I should pass t0 as argument?

Reply:
Both arguments are good arguments. To take them into account I propose `build_tableau(self)` becomes `build_tableau(self, target)`, we define the variables as `torch.tensor(alpha, dtype=target.dtype, device=target.device)` and then in `step`:
```
if self.tableau is None:
    self.tableau = self.build_tableau(f0)
```

What do you think?

Reply:
Sounds good to me !